game-core/src/main/java/games/strategy/triplea/ai/pro/data/ProTerritoryManager.java

```diff
@@ -360,103 +359,33 @@ private static void findScrambleOptions(
-            .and(Matches.unitIsNotDisabled())
-            .and(Matches.unitIsBeingTransported().negate());
-    final Predicate<Territory> canScramble =
-        PredicateBuilder.of(Matches.territoryIsWater().or(Matches.isTerritoryEnemy(player, data)))
-            .and(
-                Matches.territoryHasUnitsThatMatch(
-                    Matches.unitCanScramble()
-                        .and(Matches.unitIsEnemyOf(data, player))
-                        .and(Matches.unitIsNotDisabled())))
-            .and(Matches.territoryHasUnitsThatMatch(airbasesCanScramble))
-            .andIf(fromIslandOnly, Matches.territoryIsIsland())
-            .build();
-
-    // Find potential territories to scramble from
-    final Map<Territory, Set<Territory>> scrambleTerrs = new HashMap<>();
-    for (final Territory t : moveMap.keySet()) {
-      if (t.isWater() || !toSeaOnly) {
-        final Set<Territory> canScrambleFrom =
-            new HashSet<>(
-                CollectionUtils.getMatches(
-                    data.getMap().getNeighbors(t, maxScrambleDistance), canScramble));
-        if (!canScrambleFrom.isEmpty()) {
-          scrambleTerrs.put(t, canScrambleFrom);
-        }
-      }
-    }
-    if (scrambleTerrs.isEmpty()) {
-      return;
-    }
-
-    // Find potential max units that can be scrambled to each territory
-    for (final Territory to : scrambleTerrs.keySet()) {
-      for (final Territory from : scrambleTerrs.get(to)) {
-
-        // Find potential scramble units from territory
-        final Collection<Unit> airbases = from.getUnitCollection().getMatches(airbasesCanScramble);
-        final int maxCanScramble = getMaxScrambleCount(airbases);
-        final Route toBattleRoute =
-            data.getMap().getRoute(from, to, Matches.territoryIsNotImpassable());
-        List<Unit> canScrambleAir =
-            from.getUnitCollection()
-                .getMatches(
-                    Matches.unitIsEnemyOf(data, player)
-                        .and(Matches.unitCanScramble())
-                        .and(Matches.unitIsNotDisabled())
-                        .and(Matches.unitWasScrambled().negate())
-                        .and(Matches.unitCanScrambleOnRouteDistance(toBattleRoute)));
-
-        // Add max scramble units
-        if (maxCanScramble > 0 && !canScrambleAir.isEmpty()) {
-          if (maxCanScramble < canScrambleAir.size()) {
-            canScrambleAir.sort(
-                Comparator.<Unit>comparingDouble(
-                        o ->
-                            ProBattleUtils.estimateStrength(
-                                to, List.of(o), new ArrayList<>(), false))
-                    .reversed());
-            canScrambleAir = canScrambleAir.subList(0, maxCanScramble);
-          }
-          moveMap.get(to).getMaxScrambleUnits().addAll(canScrambleAir);
+    final var scrambleLogic = new ScrambleLogic(data, player, moveMap.keySet());
+    for (final var territoryToScramblersEntry :
```

Comment:
Would it be better to specify the type here instead of var since its not very obvious?

Reply:
Jumping in, the first var here, I'm not sure if it would help. The type is convoluted, it's mostly noise. We have an assignment in the following lines, which are probably going to be the thing that is most helpful anyways. Replacing the 2nd var IMO perhaps could be beneficial.

Reply:
Fair point and agree the second is probably more useful than the first. Mostly I just want to have some idea over what I'm iterating over as otherwise its pretty hard to follow.

Reply:
Kept var here, but added type on the inner loop.